### PR TITLE
Use `Dhall.Eval` for type-checking

### PR DESCRIPTION
**Pull request by Gabriel439**
Fixes https://github.com/dhall-lang/dhall-haskell/issues/1306

The following comment describe how this works in depth:

https://github.com/dhall-lang/dhall-haskell/issues/1129#issuecomment-532928798

The short explanation is that we can substantially speed up type-checking
by not substituting `let` expressions and instead adding the bound variable
to the context.  This is the same as the type-checking "fast path" that
we had before adding dependent types, except this time it works even in the
presence of dependent types.

The main difference is that we add the `let`-bound type *and* value to the
context and use `Dhall.Eval.conv` to perform equivalence checking instead of
`Dhall.Core.judgmentallyEqual`.

### CODE REVIEW

**Comment by Gabriel439** [comment]
Here a benchmark for the example that prompted this work:

```
$ time dhall --file ~/proj/cpkg/pkgs/pkg-set.dhall 1>/dev/null

real	0m16.792s
user	0m14.264s
sys	0m1.485s
```

And here is an example on `dhall-kubernetes`:

```
$ DHALL_PRELUDE=https://prelude.dhall-lang.org/package.dhall time dhall --file ~/proj/dhall-kubernetes/examples/deployment.dhall 1>/dev/null
        0.69 real         0.49 user         0.05 sys
```

**Comment by amarrella** [comment]
Hello!

I'm suffering a small performance hit with this version on records with both types and defaults. The impact is small but noticeable. 

On [EarnestResearch/dhall-packages](https://github.com/EarnestResearch/dhall-packages):
```sh
➜  dhall-packages git:(master) time dhall --file ./package.dhall > /dev/null
dhall --file ./package.dhall > /dev/null  4.04s user 0.25s system 82% cpu 5.205 total
➜  dhall-packages git:(master) time dhall --file ./package.dhall > /dev/null
dhall --file ./package.dhall > /dev/null  4.15s user 0.33s system 83% cpu 5.368 total
➜  dhall-packages git:(master) time dhall --file ./package.dhall > /dev/null
dhall --file ./package.dhall > /dev/null  4.17s user 0.33s system 78% cpu 5.719 total
➜  dhall-packages git:(master) time dhall --file ./package.dhall > /dev/null
dhall --file ./package.dhall > /dev/null  4.01s user 0.25s system 82% cpu 5.157 total
➜  dhall-packages git:(master) time dhall-fast-typecheck-3 --file ./package.dhall > /dev/null
dhall-fast-typecheck-3 --file ./package.dhall > /dev/null  5.07s user 0.47s system 84% cpu 6.575 total
➜  dhall-packages git:(master) time dhall-fast-typecheck-3 --file ./package.dhall > /dev/null
dhall-fast-typecheck-3 --file ./package.dhall > /dev/null  5.04s user 0.46s system 85% cpu 6.414 total
➜  dhall-packages git:(master) time dhall-fast-typecheck-3 --file ./package.dhall > /dev/null
dhall-fast-typecheck-3 --file ./package.dhall > /dev/null  5.02s user 0.47s system 85% cpu 6.380 total
➜  dhall-packages git:(master) time dhall-fast-typecheck-3 --file ./package.dhall > /dev/null
dhall-fast-typecheck-3 --file ./package.dhall > /dev/null  5.06s user 0.52s system 86% cpu 6.461 total
```

Not sure if this is important or not right now (the case in https://github.com/dhall-lang/dhall-haskell/issues/1306 is probably more important), but just wanted to highlight it

**Comment by Gabriel439** [comment]
@amarrella: Yeah, it will deteriorate performance a little in some cases, but not drastically (and there are still more optimizations we can do).  However, overall it should be a net win for almost all use cases.

For example, although `dhall-packages` might take a little longer to type-check in isolation, it now only needs to be type-checked once when used in downstream code, instead of being type-checked once every time a user accesses a field from the `dhall-packages` record.

File: dhall/src/Dhall/Eval.hs [hunk 1]
```diff
@@ -128,6 +138,11 @@ data HLamInfo a
 pattern VPrim :: (Val a -> Val a) -> Val a
 pattern VPrim f = VHLam Prim f
 
+toVHPi :: Eq a => Val a -> Maybe (Text, Val a, Val a -> Val a)
```
Comment (sjakobi) [inline dhall/src/Dhall/Eval.hs]: :+1:

**Pull request merged**